tools/fapi/tss2_template.c

```diff
@@ -250,6 +250,75 @@ TSS2_RC auth_callback(
     return TSS2_RC_SUCCESS;
 }
 
+TSS2_RC sign_callback(
+    char        const *description,
+    char        const *objectPath,
+    char        const *publicKey,
+    char        const *publicKeyHint,
+    uint32_t    hashAlg,
+    uint8_t     const *dataToSign,
+    size_t      dataToSignSize,
+    uint8_t     const **signature,
+    size_t      *signatureSize,
+    __attribute__((unused)) void *userData)
+{
+    /* The public key that will be used by the TPM to verify the signature in PEM encoding */
+    (void) publicKey; // 
+    (void) publicKeyHint;
+    (void) hashAlg;
+    (void) dataToSign;
+    (void) dataToSignSize;
+
+    int rc;
+    char *default_name = "Data2Sign.file";
+    char buf[1024];
+
+    if (userData) {
+        printf("%s:", (const char *) userData);
+    } else {
+        printf ("Authorize %s with \"%s\"\n", description, objectPath);
+    }
+    printf("Filename for nonce output: ");
+    if (buf == fgets(buf, 1024, stdin)) {
+        char path[256];
+        rc = sscanf(buf, "%s", path);
+        if (rc == -1)
+            strcpy(path, default_name);
+
+        rc = open_write_and_close(path, true, dataToSign,
+        dataToSignSize);
+        if (rc) {
+            fprintf (stderr, "Could not write to file: %s\n", path);
+            return TSS2_FAPI_RC_GENERAL_FAILURE;
+        }
+
+        printf("Filename for signature input: ");
+        if (buf == fgets(buf, 1024, stdin)) {
+            char path[256];
+            rc = sscanf(buf, "%s", path);
+            if (rc != 1){
+                fprintf (stderr, "Please enter a valid file path\n");
+                return TSS2_FAPI_RC_GENERAL_FAILURE;
+            }
```

Comment:
These fgets/sscanf blocks seem like a great helper function if we want a safer version of scanf.

Reply:
Where would be an appropriate location for this helper function? In /lib/tpm2_util.c?

Reply:
Yes.

Reply:
Done in current version